### ISSUE TITLE
knife bootstrap windows fails when winrm password contains an ampersand

### ISSUE THREAD

**Issue by donbecker** [chef/chef#5244]
## Description

Following the tutorial here: [https://learn.chef.io/manage-a-node/windows/bootstrap-your-node/](url)

I am attempting to run the following command in Powershell as Admin: 

`knife bootstrap windows winrm 52.201.225.210 --winrm-user Administrator --winrm-password 'HhrQCP&e*Ey' --node-name node1 --run-list 'recipe[learn_chef_iis]'`

Instead of bootstrapping the node, I get the following error: 

ERROR: You must pass a node name with -N when bootstrapping with user credentials.
## Chef Version

Chef DK 0.17.17
chef-client version 12.13.37
## Platform Version

Windows 7 Enterprise Service Pack 1
## Replication Case

Attempting to bootstrap a Windows node directly with user credentials fails with "You must pass.." error above. 

I have tried both --node-name and -N options to pass the node name in. I have also tried wrapping in single quotes, but still receive the same error. I also receive the same error if I omit the node name.

I have confirmed connectivity to the instance via: 

knife wsman test my.server.public.ip --manual-list

...so I am confident this is not a connectivity issue.

**Comment by donbecker** [comment]
Update: 

I have been able to get past the error by changing the order of the parameters being passed. 

Here is the order that just worked: 

knife bootstrap windows winrm my.server.public.ip --node-name node1 --winrm-user username --win-rm-password 'password' --run-list 'recipe[learn_chef_iis]'

This is bootstrapping an instance on Google Cloud, and the password is autogenerated by Google and contains / can contain special characters such as | %&*. 

The password in the above command is wrapped in single quotes, should that be sufficient to handle these characters in the password?

**Comment by donbecker** [comment]
As I mentioned, I'm past the first error, but now getting another error that may relate to the password characters: 

Creating new client for node1
Creating new node for node1
.ERROR: Failed to authenticate to my.server.public.ip as username
Response: WinRM::WinRMAuthorizationError
Hint: Make sure to prefix domain usernames with the correct domain name.
Hint: Local user names should be prefixed with computer name or IP address.

I have confirmed I can directly RDP to the machine using IP\username as the username. 

I have tried the bootstrap command with IP\username as the username, but still receive the WinRMAuthorizationError.

**Comment by donbecker** [comment]
After researching further, I think my WinRM connection issues are a separate matter not relating to Chef. 

Leaving this open due to the affect the parameter order had above, not sure if that is a chef bug or not.

**Comment by donbecker** [comment]
After trying various combinations of passwords, I've determined that passing a valid windows username & password wherein the password has one ampersand character fails.

**Comment by donbecker** [comment]
After determining the ampersand in the password caused the error, setting the node name parameter back to the typical order results in no failure, so the original error I had was related to the password characters.

**Comment by lamont-granquist** [comment]
probably something to do with the use of shell_out() and passing shell metacharacters and not handling them properly.

generally we should clean up the use of shell_out() and i noticed recently that if you use splat-args for arguments it automatically protects args against filenames-with-spaces (`shell_out('command', 'stuff', 'filename with spaces', *options)` which we should really turn into a best-practice.

**Comment by lamont-granquist** [comment]
although this may be a bug in knife-windows, but its probably also buggy elsewhere in the core chef codebase

**Comment by btm** [comment]
@donbecker `&` means something special to powershell, even in quotes. Try running`Import-Module chef` once before your knife command. Alternately you can disable powershell argument parsing for an individual command by inserting `--%`, like:

``` 
knife --% bootstrap windows winrm 52.201.225.210 --winrm-user Administrator --winrm-password 'HhrQCP&e*Ey' --node-name node1 --run-list 'recipe[learn_chef_iis]'
```

More information on this powershell module which was merged in https://github.com/chef/chef/pull/3339 is available here: https://blog.chef.io/2015/06/24/chef-client-12-4-0-released/, which was originally #1687 and https://github.com/chef/knife-windows/issues/249.

I don't think we can do more than add additional documentation callouts at this point, because it's in the shell.

The learnchef page linked by the OP is now at https://learn.chef.io/tutorials/manage-a-node/windows/hosted/bootstrap-your-node/. I'll mention adding a note to our internal learnchef team. Beyond that, I'm closing this as a duplicate of the earlier issues.
